File: typescript/src/get-max-gifts.ts
function iterateThruCities(value: number, maxLevel: number, elements: Array<number>, limit: number, count: number): number {
	if (value === limit) return value
	if (value > limit) return 0
	if (count >= maxLevel) return value

	const values: number[] = []

	for (let index in elements) {
		const newElements = [...elements]
		const currentValue = newElements.splice(Number(index), 1)[0]

		values.push(
			iterateThruCities(value + currentValue, maxLevel, newElements, limit, count + 1)
		)
	}

	return Math.max(...values)
}

function getMaxGifts(giftsCities: Array<number>, maxGifts: number, maxCities: number) {
	const currentSums: number[] = []

	for (let index in giftsCities) {
		const giftsCitiesCopy = [...giftsCities]
		const currentCity = giftsCitiesCopy.splice(Number(index), 1)[0]
		currentSums.push(
			iterateThruCities(currentCity, maxCities, giftsCitiesCopy, maxGifts, 1)
		)
	}

	const bestFit = Math.max(...currentSums)
	return bestFit > maxGifts ? 0 : bestFit
}

const giftsCities = [12, 3, 11, 5, 7]
const maxGifts = 20
const maxCities = 3


console.log(getMaxGifts([12, 3, 11, 5, 7], 20, 3)) // 20
console.log(getMaxGifts([50], 15, 1)) // 0
console.log(getMaxGifts([50], 100, 1)) // 50
console.log(getMaxGifts([50, 70], 100, 1)) // 70
console.log(getMaxGifts([50, 70, 30], 100, 2)) // 100
console.log(getMaxGifts([50, 70, 30], 100, 3)) // 100
console.log(getMaxGifts([50, 70, 30], 100, 4)) // 100

File: python/src/get_max_gifts.py
from typing import List
from functools import reduce

def iterate_thru_cities(value: int, max_level: int, elements: List[int], limit: int, count: int):
	if value == limit:
		return value

	if value > limit:
		return 0

	if count >= max_level:
		return value

	values: List[int] = []

	for index, current_value in enumerate(elements):
		new_elements = elements.copy()
		new_elements.pop(index)

		values.append(
			iterate_thru_cities(value + current_value, max_level, new_elements, limit, count + 1)
		)

	return reduce(lambda a, b: a if a > b else b, values)


def get_max_gifts(gifts_cities: List[int], max_gifts: int, max_cities: int):
	current_sums: List[int] = []

	for index, current_city in enumerate(gifts_cities):
		gifts_cities_copy = gifts_cities.copy()
		gifts_cities_copy.pop(index)

		current_sums.append(
			iterate_thru_cities(current_city, max_cities, gifts_cities_copy, max_gifts, 1)
		)

	best_fit = reduce(lambda a, b: a if a > b else b, current_sums)
	return 0 if best_fit > max_gifts else best_fit

if __name__ == "__main__":
	print(get_max_gifts([12, 3, 11, 5, 7], 20, 3)) # 20
	print(get_max_gifts([50, 70, 30], 100, 2)) # 100
	print(get_max_gifts([50, 70, 30], 100, 3)) # 100

	print(get_max_gifts([50], 15, 1)) # 0
	print(get_max_gifts([50], 100, 1)) # 50
	print(get_max_gifts([50, 70], 100, 1)) # 70
	print(get_max_gifts([50, 70, 30], 100, 4)) # 100